middlewares/retry.go

```diff
@@ -32,12 +33,19 @@ func NewRetry(attempts int, next http.Handler) *Retry {
 }
 
 func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	// if we might make multiple attempts, swap the body for an ioutil.NopCloser
+	// cf https://github.com/containous/traefik/issues/1008
+	if retry.attempts > 1 {
+		body := r.Body
+		defer body.Close()
+		r.Body = ioutil.NopCloser(body)
+	}
 	attempts := 1
 	for {
 		recorder := NewRecorder()
 		recorder.responseWriter = rw
 		retry.next.ServeHTTP(recorder, r)
-		if !isNetworkError(recorder.Code) || attempts >= retry.attempts {
+		if recorder.Error == nil || attempts >= retry.attempts {
```

Comment:
I've reverted it, however please note that it wouldn't change the behaviour for errors you're mentioning, they'd still lead to retries.

The errors I was tackling with this change are instead explicit http errors from upstream, to compare this according to [proxy_next_upstream](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream) semantics, currently the behaviour is : 

    error timeout http_502 http_504 non_idempotent

And with my changes it would be : 

    error timeout non_idempotent

The motivation behind this is that imo `http_5XX + non_idempotent` is a bit aggressive for a default. If upstream sends http 504 for a POST request it still might have altered a resource but didn't manage to serve the response in a timely manner, and retrying might not be the best option, at least hard to assume. Maybe even `non_idempotent` shouldn't be by default? Or the retry behaviour should be configurable?